### ISSUE TITLE
Firestore conformance test for Listen parentPath

### ISSUE THREAD

**Issue by jadekler** [googleapis/google-cloud-common#273]
In switching to v1, firestore clients need to append /documents to parent paths. Several RPCs will fail if this does not happen, which illustrates the need for /documents, but _not all RPCs will fail_. In particuar, the QueryTarget.Parent part of the Listen feature needs this but the proxy will not check for it (yet).

We should add a conform test that makes sure all client libraries are building these paths correctly.

However, our [test.proto is not set up to do so](https://github.com/googleapis/google-cloud-common/blob/master/testing/firestore/proto/test.proto#L161-L176) *, and therefore neither are our test runners.

* Our listen tests currently function as such: given some response from firestore, some snapshot should be created. What we need for this issue is a test such that given some API call by the user constructs a proto in some specific way (in this issue we care that it has a specific parent path).

**Comment by jadekler** [comment]
cc @mikelehen @schmidt-sebastian @jskeet

**Comment by tseaver** [comment]
Rather than come up with some dance around the already-brittle conformance tests, couldn't we just add / update "normal" unit tests for the V1 calls, asserting that the correct path is passed to the API?

**Comment by schmidt-sebastian** [comment]
The part of the proto that this change touches (such as https://cs.corp.google.com/piper///depot/google3/google/firestore/v1beta1/firestore.proto?dr=CSs&g=0&l=552) has so far not been part of the conformance test suite. If we have spare cycles, I would rather steer them in a different direction and add more tests for retry and resume behavior in the Watch API.

Not including the `/documents` suffix should break system tests right away, unless there are clients that never query the main collection.

**Comment by tseaver** [comment]
@jadekler Running system tests against my [`V1` PR](https://github.com/googleapis/google-cloud-python/pull/7494), the "listen" tests raise the following error, which suggests that a different fix than appending `documents`:

```
400 Missing required http header 'google-cloud-resource-prefix' or query param 'database'.
```

Can you point to where the V1 API docs describe the new requirement?  The API docs for `ListenRequest` are identical between [`v1beta1`](https://cloud.google.com/firestore/docs/reference/rpc/google.firestore.v1beta1#google.firestore.v1beta1.ListenRequest) and [`v1`](https://cloud.google.com/firestore/docs/reference/rpc/google.firestore.v1#google.firestore.v1.ListenRequest).

**Comment by schmidt-sebastian** [comment]
This has been a requirement for a while now, but it wasn't enforced: https://docs.google.com/document/d/1PjoomZav75_wevZxTarHWqkAlcLnNBAhLBJ4NN_eCPs/edit#bookmark=id.ir3srdzadcgm

This needs to be set as a GRPC Header.

**Comment by tseaver** [comment]
@schmidt-sebastian Why isn't the requirement spelled out in the public API docs?

**Comment by tseaver** [comment]
The Python client already adds that header for all the non-BIDI API calls.  @crwilcox I might need your help figuring out how to jam that header through into the `ResumableBidiRpc` bit for the listen APIs.

**Comment by schmidt-sebastian** [comment]
Paraphrasing someone (who shall not be named) from the backend team: The Watch API is at this point not meant to be externally consumed and as such we don't have public documentation that outlines all requirements.

**Comment by tseaver** [comment]
Looking at that doc, that header is supposed to be required *all* API requests, not just `Listen` / Watch.

**Comment by schmidt-sebastian** [comment]
That is correct, as the header helps with routing. I'll ping the backend team to see if it is documented anywhere.

**Comment by mickeyreiss** [comment]
I'm sure folks are already aware, but I do want highlight that this issue impacts the generated Go API library, `firestore/apiv1`, as well.

The public [`Listen` method](https://github.com/googleapis/google-cloud-go/blob/v0.44.3/firestore/apiv1/firestore_client.go#L368-L382) for a [seemingly well documented API](https://firebase.google.com/docs/firestore/reference/rpc/google.firestore.v1#google.firestore.v1.Firestore.Listen) does not work out of the box, and it is difficult to figure out the required incantation without access to the internal documentation referenced above.

**Comment by schmidt-sebastian** [comment]
@mickeyreiss At this point, we don't expect external customers to use these endpoints without going through one of our official clients. The protocol handles a number of state transitions that are important to the protocol but difficult to get right. If one of our existing implementations does not fit your use case, I would advocate that you them as a base and start your journey from there.

We might publish some more extensive documentation at a later point that will make it easier for third-party customers to build their own Listen client, but for now please take a look at one of the seven clients that we released.
